### DISCUSSION TITLE
Nested write issue

### DISCUSSION THREAD

**Discussion by riccardolardi** [prisma/prisma#18877] [Q&A] [answered]
I get a TS error (and failing query) when trying to execute this nested write.

```prisma
model Applicant {
  id           String        @id @default(uuid()) @db.Uuid
  email        String        @unique
  applications Application[]

  @@map("applicants")
}

model Application {
  id          String     @id @default(uuid()) @db.Uuid
  name        String
  applicant   Applicant? @relation(fields: [applicantId], references: [id])
  applicantId String?    @map("applicant_id") @db.Uuid

  @@index([applicantId])
  @@map("applications")
}
```

```js
prisma.application.create({
    data: { // Types of property 'name' are incompatible. Type 'string' is not assignable to type 'undefined'
      name: "whatever",
      applicant: {
        create: {
          email: "some@email.com",
        },
      },
    },
  })
```

What is going on here? Why is this not a valid nested write?

When executing the error throws: `Unknown arg "applicant" in data.applicant for type ApplicationUncheckedCreateInput.`

**Comment by ludralph** [comment]
Hi @riccardolardi 👋 

I'm unable to reproduce this as the nested write works correctly for me. Can you please share a repository so i can debug this further?

**Comment by riccardolardi** [comment]
Hi @ludralph thanks for looking into this.

I'm sorry I can't share the whole repo right but I can share the schema and the tRPC router with you, maybe I am doing something wrong in the schema:

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider     = "postgresql"
  url          = env("DATABASE_URL")
  directUrl    = env("DIRECT_URL")
  relationMode = "prisma"
}

model Organization {
  id         String    @id @default(uuid()) @db.Uuid
  created_at DateTime  @default(now()) @db.Timestamptz(6)
  updated_at DateTime? @updatedAt @db.Timestamptz(6)
  name       String?
  address    String?
  forms      Form[]
  profiles   Profile[]

  @@map("organizations")
}

model Profile {
  id             String        @id(map: "users_pkey") @db.Uuid
  created_at     DateTime      @default(now()) @db.Timestamptz(6)
  updated_at     DateTime?     @updatedAt @db.Timestamptz(6)
  email          String?       @unique(map: "users_email_key")
  fullname       String?
  phone          String?
  passwordSet    Boolean?      @default(false) @map("password_set")
  role           Role          @default(USER)
  organization   Organization? @relation(fields: [organizationId], references: [id])
  organizationId String?       @db.Uuid

  @@index([organizationId])
  @@map("profiles")
}

enum Role {
  USER
  MEMBER
  OWNER
  ADMIN
}

model Applicant {
  id           String        @id @default(uuid()) @db.Uuid
  created_at   DateTime      @default(now()) @db.Timestamptz(6)
  updated_at   DateTime?     @updatedAt @db.Timestamptz(6)
  email        String        @unique
  fullname     String?
  address      String?
  phone        String?
  applications Application[]

  @@map("applicants")
}

model Application {
  id          String     @id @default(uuid()) @db.Uuid
  created_at  DateTime   @default(now()) @db.Timestamptz(6)
  updated_at  DateTime?  @updatedAt @db.Timestamptz(6)
  form        Form       @relation(fields: [formId], references: [id])
  formId      String     @map("form_id") @db.Uuid
  applicant   Applicant? @relation(fields: [applicantId], references: [id])
  applicantId String?    @map("applicant_id") @db.Uuid
  title       String?
  content     Json?

  @@index([formId])
  @@index([applicantId])
  @@map("applications")
}

model Form {
  id             String        @id @default(uuid()) @db.Uuid
  created_at     DateTime      @default(now()) @db.Timestamptz(6)
  updated_at     DateTime?     @updatedAt @db.Timestamptz(6)
  name           String?
  apiKey         String        @unique @map("api_key")
  organization   Organization  @relation(fields: [organizationId], references: [id])
  organizationId String        @map("organization_id") @db.Uuid
  applications   Application[]

  @@index([organizationId])
  @@map("forms")
}
```

```ts
  create: publicProcedure
    .input(z.object({
      formId: z.string().uuid(),
      applicantId: z.string().uuid().optional(),
      content: z.object({
        "application-title": z.string(),
        "applicant-fullname": z.string(),
        "applicant-email": z.string().email(),
        "applicant-address": z.string(),
        "applicant-phone": z.string(),
      }),
    }))
    .mutation(async ({ ctx, input }) => {
      return await ctx.prisma.application.create({
        data: { // Types of property 'name' are incompatible. Type 'string' is not assignable to type 'undefined'
          formId: input.formId,
          applicantId: input.applicantId || applicant?.id,
          title: input.content["application-title"],
          content: input.content,
          applicant: {
            create: {
              fullname: input.content["applicant-fullname"],
              email: input.content["applicant-email"],
              address: input.content["applicant-address"],
              phone: input.content["applicant-phone"],
            },
          },
        },
      });
    }),
```

**Reply by ludralph** [reply]
Thank you for sharing. I'm taking a look and will get back to you

**Reply by ludralph** [reply]
Hi @riccardolardi, what are the fields allowed for the type `ApplicationUncheckedCreateInput` in /node_modules/.prisma/client/index.d.ts. That should give you a clue of the allowable fields for the application query.

**Comment by riccardolardi** [comment]
Hi Raphael

The type of ApplicationUncheckedCreateInput is:

export type ApplicationUncheckedCreateInput = {
 id?: string
 created_at?: Date | string
 updated_at?: Date | string | null
 formId: string
 applicantId?: string | null
 title?: string | null
 content?: NullableJsonNullValueInput | InputJsonValue
 }

I don’t understand why there is no “applicant” field, according to the schema (https://github.com/prisma/prisma/discussions/18877#discussioncomment-5705194) it should be possible to do a nested write, no?

Thanks, best regards
Riccardo

–
Riccardo Lardi
Software Engineering & Interaction Design
www.riccardolardi.com
***@***.*** / +41 79 339 65 38
LinkedIn / GitHub / Twitter

**Reply by ludralph** [reply, marked as answer]
Hi @riccardolardi 

When performing a `create` operation, data is the type of `XOR<ApplicationCreateInput,ApplicationUncheckedCreateInput >` which means you can either provide ApplicationCreateInput which will have all nested relations specified or you can directly provide the ID as in the case of `applicantId` field but you can’t mix them. So you will need to use the fields specified in `ApplicationUncheckedCreateInput` in order to create it. Also, `Unchecked` input types allow you to perform some operations that Prisma considers dangerous like directly writing foreign keys. Prisma allows you to choose either a safe or an unchecked input type when doing operations like create. In this case, Prisma wants you to perform the create action using the `Unchecked` input type.

**Comment by riccardolardi** [comment]
Thanks a lot Raphael, that was it. It works now.

Great support!

All best,
Riccardo

–
Riccardo Lardi
Software Engineering & Interaction Design
www.riccardolardi.com
***@***.*** / +41 79 339 65 38
LinkedIn / GitHub / Twitter

**Reply by ludralph** [reply]
You're welcome @riccardolardi. Don't hesitate to open a new discussion if you run into any issues.

**Comment by igokul1973** [comment]
I came across the same problem a few times, so in order to avoid the confusion I usually create a separate `data` variable with specifically assigned type and fill it up until I can pass it to create/update/etc. method.
For example, when I know that I want to `create` (or `connect`) instead of assigning IDs, I pick `ApplicationCreateInput` out of XOR<ApplicationCreateInput,ApplicationUncheckedCreateInput> type and assign it to `data` variable. This way I know, assuming for example, that my `Application` prisma model has `userId` and related `user` object, assignment of type `ApplicationCreateInput` to variable `data` will make sure that you do not pass the `userId` when trying to create/connect to/delete/etc. nested `user` object.
```
... some previous logic

let data: ApplicationCreateInput = {}; // deliberate assignment of the right data shape type

... some logic probably preparing `someData` and `someUserObjectWithoutId`...

data = {
   ...someData, // if I pass `userId` here, the IDE/typescript will complain that it should not exist
   user: {
        create: {
           ...someUserObjectWithoutId
        }
    }
};
prisma.application.update({
    data, 
    where: { id: app.id }}
)...
```
